src/diffusers/models/transformers/transformer_sd3.py

```diff
@@ -330,6 +331,77 @@ def _set_gradient_checkpointing(self, module, value=False):
         if hasattr(module, "gradient_checkpointing"):
             module.gradient_checkpointing = value
 
+    def _load_ip_adapter_weights(self, state_dict: Dict, low_cpu_mem_usage: bool):
+        # IP-Adapter cross attention parameters
+        hidden_size = self.config.attention_head_dim * self.config.num_attention_heads
+        ip_hidden_states_dim = self.config.attention_head_dim * self.config.num_attention_heads
+        timesteps_emb_dim = state_dict["ip_adapter"]["0.norm_ip.linear.weight"].shape[1]
+
+        # Dict where key is transformer layer index, value is attention processor's state dict
+        # ip_adapter state dict keys example: "0.norm_ip.linear.weight"
+        layer_state_dict = {idx: {} for idx in range(len(self.attn_processors))}
+        for key, weights in state_dict["ip_adapter"].items():
+            idx, name = key.split(".", maxsplit=1)
+            layer_state_dict[int(idx)][name] = weights
+
+        # Create IP-Adapter attention processor
+        attn_procs = {}
+        for idx, name in enumerate(self.attn_processors.keys()):
+            attn_procs[name] = IPAdapterJointAttnProcessor2_0(
+                hidden_size=hidden_size,
+                ip_hidden_states_dim=ip_hidden_states_dim,
+                head_dim=self.config.attention_head_dim,
+                timesteps_emb_dim=timesteps_emb_dim,
+            ).to(self.device, dtype=self.dtype)
+
+            if not low_cpu_mem_usage:
+                attn_procs[name].load_state_dict(layer_state_dict[idx], strict=True)
+            else:
+                load_model_dict_into_meta(
+                    attn_procs[name], layer_state_dict[idx], device=self.device, dtype=self.dtype
+                )
+
+        self.set_attn_processor(attn_procs)
+
+        # Convert image_proj state dict to diffusers
+        image_proj_state_dict = {}
+        for key, value in state_dict["image_proj"].items():
+            if key.startswith("layers."):
+                idx = key.split(".")[1]
+                key = key.replace(f"layers.{idx}.0.norm1", f"layers.{idx}.ln0")
+                key = key.replace(f"layers.{idx}.0.norm2", f"layers.{idx}.ln1")
+                key = key.replace(f"layers.{idx}.0.to_q", f"layers.{idx}.attn.to_q")
+                key = key.replace(f"layers.{idx}.0.to_kv", f"layers.{idx}.attn.to_kv")
+                key = key.replace(f"layers.{idx}.0.to_out", f"layers.{idx}.attn.to_out.0")
+                key = key.replace(f"layers.{idx}.1.0", f"layers.{idx}.adaln_norm")
+                key = key.replace(f"layers.{idx}.1.1", f"layers.{idx}.ff.net.0.proj")
+                key = key.replace(f"layers.{idx}.1.3", f"layers.{idx}.ff.net.2")
+                key = key.replace(f"layers.{idx}.2.1", f"layers.{idx}.adaln_proj")
+            image_proj_state_dict[key] = value
```

Comment:
Should we update the checkpoint [InstantX/SD3.5-Large-IP-Adapter](https://huggingface.co/InstantX/SD3.5-Large-IP-Adapter) (convert to safetensors while we're at it?). Otherwise we might need some extra checks here

Reply:
that's a good idea!

Reply:
This should do it, I will upload the checkpoint and propose the change.

```python
from safetensors.torch import save_file
from diffusers.utils import _get_model_file
from diffusers.models.modeling_utils import load_state_dict

# Get weights from the hub
model_file = _get_model_file(
    "InstantX/SD3.5-Large-IP-Adapter",
    weights_name="ip-adapter.bin",
)
state_dict = load_state_dict(model_file)

# ip_adapter stays the same
safetensors_dict = {f"ip_adapter.{key}": value for key, value in state_dict["ip_adapter"].items()}

# image_proj needs some adjustments
for key, value in state_dict["image_proj"].items():
    if key.startswith("layers."):
        idx = key.split(".")[1]
        key = key.replace(f"layers.{idx}.0.norm1", f"layers.{idx}.ln0")
        key = key.replace(f"layers.{idx}.0.norm2", f"layers.{idx}.ln1")
        key = key.replace(f"layers.{idx}.0.to_q", f"layers.{idx}.attn.to_q")
        key = key.replace(f"layers.{idx}.0.to_kv", f"layers.{idx}.attn.to_kv")
        key = key.replace(f"layers.{idx}.0.to_out", f"layers.{idx}.attn.to_out.0")
        key = key.replace(f"layers.{idx}.1.0", f"layers.{idx}.adaln_norm")
        key = key.replace(f"layers.{idx}.1.1", f"layers.{idx}.ff.net.0.proj")
        key = key.replace(f"layers.{idx}.1.3", f"layers.{idx}.ff.net.2")
        key = key.replace(f"layers.{idx}.2.1", f"layers.{idx}.adaln_proj")
    safetensors_dict[f"image_proj.{key}"] = value

# export!
save_file(safetensors_dict, "ip_adapter.safetensors")
```

Reply:
PR with updated checkpoints open [here](https://huggingface.co/InstantX/SD3.5-Large-IP-Adapter/discussions/10), adding safetensors as well.